### ISSUE TITLE
Pretest fails on win10

### ISSUE THREAD

**Issue by Nervengift** [veracrypt/VeraCrypt#160]
I tried to crypt a bunch of different cheap laptops (11", Intel Atom processor, small SSD) using veracrypt (1.19). All of them are running Windows 10.

On two of them (one is a lenovo ideapad 100S) the pretest failed with an error message saying "Invalid Parameter. Source: VeraCrypt::Elevator::ReadWriteFile:174"

I don't really know about debugging such stuff on Windows (I'm more of a Linux guy) but here's a procmon logfile of me trying to run the pretest: https://files.nerven.gift/f/37cd8cd07a/?dl=1

It says something about `INVALID PARAMETER` during a `QueryAttributeTagFile` on different files from `\Device\HarddiskVolume1\EFI\`

If you tell me what additional information you need and how to retreive it under Windows I'll gladly provide it.

**Comment by andrey-che** [comment]
I confirm. I've got a Prestigio Smartbook 141A03 with Intel Atom and SSD.
Totaly the same issue, on pretest I get "VeraCrypt::Elevator::ReadWriteFile:174"

P.S. Security boot turnd off

**Comment by mateuszbrylski** [comment]
I have the same problem.

**Comment by deftomat** [comment]
The same for me. Did anyone find out how to fix this annoying issue? 😞

**Comment by Nervengift** [comment]
Sadly no. And I don't know anybody with enough Windows knowledge to help me debug this.

I can only offer my time for testing stuff and wait for a developer's response…

**Comment by idrassi** [comment]
I tried to analyze this issue following other reports on Sourceforge and I started implementing some workaround but it doesn't seem to work.

In version 1.21 that was just released, I added debugging symbols so that it can be easier to analyze the problem.
@Nervengift  Can you please test again using 1.21 and create a new procmon log file? This will help a lot in understanding what is causing the issue. 

Also, it will be helpful to do the following simple test: open a command prompt as an administrator using mouse right click and then type the following command sequence:
- `mountvol.exe B: /S`
- `B:`
- `cd EFI`
- `mkdir VeraCrypt`
- `cd VeraCrypt`
- `echo "test" > "test.txt"`
- `more test.txt`
- `del test.txt`
- `cd ..`
- `rmdir VeraCrypt`

Do you have any errors when performing this sequence?

One thing is clear: something in these machines is not allowing VeraCrypt to install its bootloader files in the EFI system partition although we have full admin privileges. Either these machines require some special system calls or the EFI system partition is completely protected against modification.

**Comment by deftomat** [comment]
@Nervengift I tried your command sequence and there was no errors. Everything works as expected.

**Comment by deftomat** [comment]
@Nervengift I tried your command sequence and there was no errors. Everyhing works as expected.

**Comment by idrassi** [comment]
@deftomat This means that the way VeraCrypt writes data to EFI system partition is not compatible with your system. This is strange and requires some testing. @Nervengift procmon logs can help see what is going on under the hood.

Last question: are you running a 32-bit version of Windows 10 or a 64-bit one? @Nervengift logs show that he is running 32-bit Windows so maybe this happens only on 32-bit system.

**Comment by deftomat** [comment]
@idrassi Yes, it is a 32-bit version.

**Comment by Nervengift** [comment]
@idrassi here's the new log file: https://files.nerven.gift/f/c9d87ccce7/?dl=1 . The error dialog shows exactly the same error as before but I guess that was to be expected?

The command sequence works without errors for me as well

**Comment by idrassi** [comment]
After deeper analysis, it looks like the issue is caused by the characteristics of the hard drive used by the machine.
Can you please open the command prompt as an administrator and type the following command:
- `fsutil fsinfo ntfsinfo c:`

and post the value displayed in the lines "Bytes Per Sector" and "Bytes Per Physical Sector"?

One possibility is that "Bytes Per Sector" has a value different from the expected 512 value (probably 4096). 
For now, VeraCrypt expected the disk where system encryption takes place to have a sector size of 512 bytes. Supporting 4096 bytes value requires some changes that I already started.

**Comment by deftomat** [comment]
Result of a `fsutil fsinfo ntfsinfo c:`

```
NTFS Volume Serial Number :        0x28bc27bbbc278284
NTFS Version   :                   3.1
LFS Version    :                   2.0
Number Sectors :                   0x00000000037befff
Total Clusters :                   0x00000000006f7dff
Free Clusters  :                   0x00000000000faed6
Total Reserved :                   0x000000000000320d
Bytes Per Sector  :                512
Bytes Per Physical Sector :        4096
Bytes Per Cluster :                4096
Bytes Per FileRecord Segment    :  1024
Clusters Per FileRecord Segment :  0
Mft Valid Data Length :            0x000000000ff80000
Mft Start Lcn  :                   0x00000000000c0000
Mft2 Start Lcn :                   0x0000000000000002
Mft Zone Start :                   0x00000000003bae80
Mft Zone End   :                   0x00000000003c2240
Max Device Trim Extent Count :     1
Max Device Trim Byte Count :       0x8000000
Max Volume Trim Extent Count :     1
Max Volume Trim Byte Count :       0x8000000
Resource Manager Identifier :     2D76EBA5-67D2-11E6-BC55-000EAA005187
```

**Comment by idrassi** [comment]
@deftomat Thanks. So the sector size is not the culprit.
To help me more, can you please install version 1.21 and then start "VeraCrypt Format.exe" as an administrator using mouse right click and in the wizard, choose "Encrypt the system...". Then please post the error message with method/line number.
Thank you.

**Comment by deftomat** [comment]
@idrassi

```
The parameter is incorrect.

Source: VeraCrypt::File::Write:603
```

**Comment by idrassi** [comment]
Thank you.
I have implemented a workaround that should should solve this issue and I have uploaded a Windows installer for 1.22-BETA0 that includes it to https://sourceforge.net/projects/veracrypt/files/VeraCrypt%20Nightly%20Builds/
Can you please test it and report the result?

Basically, it looks that the disk used by Windows doesn't support using the flag `FILE_FLAG_WRITE_THROUGH` in `CreateFile` function. The workaround consists of detecting the invalid parameter error and reopen the disk without this flag.

**Comment by deftomat** [comment]
Thanks. However, it throws the same error but on a line 620.

**Comment by idrassi** [comment]
Well, the only explanation left is that there is an alignment issue although the disk reports a sector size of 512.
I have done a last resort modification that should fix the issue if it is linked to alignment.
Can you please test using the installer at https://veracrypt.fr/downloads/1.22-TEST/VeraCrypt%20Setup%201.22-BETA1-TEST.exe (pgp signature at https://veracrypt.fr/downloads/1.22-TEST/VeraCrypt%20Setup%201.22-BETA1-TEST.exe.sig)?

Also, what is the type/model of your system hard drive?

Thank you for your help.

**Comment by deftomat** [comment]
The same error at line 649.

SSD is Samsung BGND3R 32GB in Lenovo 100s.

**Comment by idrassi** [comment]
Actually this is good news!!
This specific line is executed after successful write of volume header sector on the disk and it had a wrong test. 
This means that the issue is really caused by the disk rejecting 512 bytes sectors write operations although it indicated that it is supported. Clearly this is caused by an issue in the SSD driver.

I have put a new installer with the complete fix at https://veracrypt.fr/downloads/1.22-TEST/VeraCrypt%20Setup%201.22-BETA1-TEST2.exe (pgp signature at https://veracrypt.fr/downloads/1.22-TEST/VeraCrypt%20Setup%201.22-BETA1-TEST2.exe.sig).  With this, you should be able to complete the encryption process and start the Pre-Test (be sure to disable Secure Boot from BIOS).

Thank you for your help.

**Comment by deftomat** [comment]
Ok, no errors for this time. However, after an reboot, it hangs on an `Authorizing...`. I can cancel a pre-test with ESC key and boot to Windows.

Pre-test cannot be uninstalled due to error in `VeraCrypt::Elevator::ReadWriteFile:174` (the same as when you try to encrypt a drive from a VeraCrypt software itself - not a VeraCrypt Format.exe)

I try to wait for a 10 minutes while it try to authorize. Password is less then 20chars and PIM has a default value. This PC is really slow. It tooks 1-2 minutes to mount an encrypted file with the same password and PIM. I tried to use longer password and PIM = 1. It tooks just a few seconds. 

So, I tried to encrypt drive with this longer password and PIM = 1 and it still hangs on `Authorizing...`
However, I don't know if my second try to encrypt a drive installs a valid pre-test because old pre-test was not uninstalled 😞

**Comment by idrassi** [comment]
We have two issues here.
Concernining hang on Authorizing... it is probably a slowness issue. To be sure, you can use benchmark utility to have an estimate of the time needed for password verification: go to menu "Tools -> Benchmark" and choose "PKCS-5 PRF" in the combo box, check "Pre-Boot" and type the value of PIM you wan to use (or leave empty for default). After clicking Benchmark button, it will display the estimated time needed for Pre-Boot by each hash algorithm.
Did you choose default algorithm values (AES-SHA512)?

Concerning the error to uninstall Pre-Test, can you please quit running instance of VeraCrypt and then launch VeraCrypt.exe as an administrator (right click menu) and proceed to Pre-Test uninstall: the displayed error message will be more helpful to me to find where the error occurs.

Thank you again for your previous help. Nothing happens to the drive when the Pre-Test fails so there is no risk and it is always possible to uninstall Pre-Test files manually as a last resort if automatic cleanup fails.

**Comment by deftomat** [comment]
Yes, benchmarks with a default PIM tooks forever.

Error in pre-test uninstall: 
```
The parameter is incorrect

Source: VeraCrypt::File::Read:534
```

**Comment by Nervengift** [comment]
Exactly the same behaviour here. But thats not really surprising since it's the same model (will probably try on the other ones next week)

it's really not usable with standard PIM, let's hope I can convince the guy I'm crypting those laptops for to use a >20 char password ^^"

**Comment by idrassi** [comment]
We are getting closer to solve this issue!
I have applied the same previous fix for write to the read operations and now the uninstall should be OK.
Can you please try with the new installer at https://veracrypt.fr/downloads/1.22-TEST/VeraCrypt%20Setup%201.22-BETA1-Build3.exe (pgp signature at https://veracrypt.fr/downloads/1.22-TEST/VeraCrypt%20Setup%201.22-BETA1-Build3.exe.sig)?

**Comment by Nervengift** [comment]
in both versions uninstalling fails in the autostart popup that tells me about the failed pre-test but seems to work when I start veracrypt as administrator and try again (that was the case in BETA2, too).

Also I can' get the authentication to work. Neighter short password + PIM 485 nor long password + PIM 1 showed anything but "Authenticating..." for mor than 15 minutes (short pw I even tried for > 1 hour). I'm not sure that this is only a problem with computational power, the benchmark took only ~800 ms for PIM 1 and ~30 s for PIM 485 (HMAC-SHA-512)

**Comment by idrassi** [comment]
OK, thanks for the feedback.
The EFI bootloadere must have the same issue accessing the disk to read the volume header. I will try to see if I can solve this since this is the real blocking problem now.
For the uninstall issue, this is just privilege issue that can be fixed since running as administrator solved it.

I will let you know about my advancements.

**Comment by nursoda** [comment]
I have the same issue on an Odys W10 tablet. I did not try to install TEST versions since idrassi knows even there's an issue to be solved first. Posting here to add some of the requested results and to help testing if necessary.

CPU ist Atom Z3735F, SSD is Toshiba model 032GE4. fsinfo yields 512 Bytes Per Sector and 4096 Bytes Per Physical Sector. PKCS-5 PRF Benchmark with default PIM took ~2 minutes, yields 29232ms for HMAC-SHA-512 on default 500000 iterations.

**Comment by pepawel** [comment]
Same issue with Lenovo Ideapad 100.

**Comment by kiddi3** [comment]
hello so i deferred decryption then restarted system but my system wont boot tried decrypting twice but still no show tried mounting using veracrypt on my system booting with linux but it gives (The process t encryption of the partition/drive has not been completed. It must be completed first) even after decrypting twice i dont want to loose my files please any help and i also deleted my efi 100mb partition a recomendation from someone which made my system unbootable but i guess if i can decrypt then i can fix with startup repair

**Comment by hoffepoffe** [comment]
@idrassi: I have a similar problem, maybe/probably related. I have a new, cheap MPMAN NT11 notebook with Intel Atom Z3000-series. Windows 10 32 bit. Harddrive is a 30 GB Kingston T52732. Tried running system encryption with Veracrypt 1.21, but pre-test failed with the same error (i.e. pre-test didn't even start). Installed VC 1.22 BETA3. Then it first works, but when clicking restart, the computer instead shuts down. I turn it on, it boots and shows BIOS (UEFI) test, and then suddenly the computer restarts. This time the password prompt comes, but looks weird (see pic attached). I have VC running on a normal computer as well, and there the screen is first cleared and VC shows some info before the prompt. Then I enter the correct password and PIM, which is 1 (with long password). Should take a second. Waited a few minutes and nothing happens,, so it obviously hangs at Authorizing... (see pic).

UEFI BIOS is InsydeH2O. I entered it and started Boot Manager, which shows both VC and Windows. I don't know if this is related to the problem.

Let me know if you can help, and if you want some more info.

Thanks.

![img_2847](https://user-images.githubusercontent.com/33122660/32680457-0397500e-c66b-11e7-93f4-fe2c26bb576e.jpg)

![img_2848](https://user-images.githubusercontent.com/33122660/32680463-0d621bc8-c66b-11e7-9f6b-8794b525a1ae.jpg)

**Comment by benjones201244** [comment]
Having the same issue here. Toshiba Satellite Click Mini (L9W-B), another mini-netbook using 32 GB eMMC storage. I was using Bitlocker previously but prefer VeraCrypt now. Unfortunately getting the same error others report:

_The parameter is incorrect.
Source: VeraCrypt::Elevator::ReadWriteFile:174_

Will be very happy when this is solved. Until then I will just use Bitlocker again.

**Comment by ploink** [comment]
Same problem here, _VeraCrypt::Elevator::ReadWriteFile:174_
Asus Transformer T100TA 32bit Win10 with 64GB emmc.

**Comment by thefifthuser** [comment]
The Asus X205TA is affected by the same problem.

It's also using 32 GB eMMC storage.

I think this means that VeraCrypt has a problem working on quite a wide range of netbooks.

**Comment by thefifthuser** [comment]
Update:

I tried [VeraCrypt Setup 1.22-BETA6.exe](https://sourceforge.net/projects/veracrypt/files/VeraCrypt%20Nightly%20Builds/) and it seems to work on my system.

I suppose this will be fixed in the 1.22 release then?

**Comment by ploink** [comment]
I think I just found the source op the problem! 
The EFI partition and the C drive have DIFFERENT sector properties, even though they are both on the same eMMC SSD. 

Try this in an administrator prompt and look at the **FileSystemEffectivePhysicalBytesPerSectorForAtomicity** attribute:
```
mountvol.exe B: /S

fsutil fsinfo sectorInfo c:

LogicalBytesPerSector :                                 512
PhysicalBytesPerSectorForAtomicity :                    4096
PhysicalBytesPerSectorForPerformance :                  4096
FileSystemEffectivePhysicalBytesPerSectorForAtomicity : 512
Device Alignment :                                      Aligned (0x000)
Partition alignment on device :                         Aligned (0x000)
No Seek Penalty
Trim Supported
Not DAX capable
```

```
fsutil fsinfo sectorInfo b:

LogicalBytesPerSector :                                 512
PhysicalBytesPerSectorForAtomicity :                    4096
PhysicalBytesPerSectorForPerformance :                  4096
FileSystemEffectivePhysicalBytesPerSectorForAtomicity : 4096
Device Alignment :                                      Aligned (0x000)
Partition alignment on device :                         Aligned (0x000)
No Seek Penalty
Trim Supported
Not DAX capable
```

**Comment by thefifthuser** [comment]
@ploink Have you tried any of the newer beta versions for VeraCrypt 1.22?

>Better fix for "Parameter Incorrect" issues during EFI system encryption in some machines.

https://www.veracrypt.fr/en/Release%20Notes.html

https://github.com/veracrypt/VeraCrypt/commit/3021745f67164c2a19cafc0ed1eff7d35830f662

It seems this was added shortly after the release of VeraCrypt 1.21, so it's not in there and only in the 1.22 beta versions at the moment.

Please correct me if I'm wrong.

**Comment by Duzga** [comment]
Before restart for test open Admin command promt and type:
bcdedit /set "{bootmgr}" path \EFI\Boot\bootx64.efi

**Comment by squantrill** [comment]
I had similar issues with Windows 10 home, I needed to activate the hidden administrator account, then re-login and run install again, now its working fine by the l looks of it..
